### PR TITLE
Audit log improvement

### PR DESCRIPTION
**Pull request by marcocaberletti**
This PR implements an event based approach for issue #79 .
The main changes are:
- definition of custom IAM events;
- add event listeners that log the catched events;
- modify the main IAM services to publish interesting events;
- add some unit tests.

### CODE REVIEW

File: iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthenticationAuditListener.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    implements ApplicationListener<AbstractAuthenticationEvent> {
+
+  private static final Log logger = LogFactory.getLog(IamAuthenticationAuditListener.class);
+
+  private Map<String, Object> data = Maps.newLinkedHashMap();
+
+  @Override
+  public void onApplicationEvent(AbstractAuthenticationEvent event) {
+
+    data = Maps.newLinkedHashMap();
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthenticationAuditListener.java]: Probably is more efficient to just clear the existing map, instead of allocating a new map for each event

File: iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthorizationAuditListener.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  private static final Log logger = LogFactory.getLog(IamAuthorizationAuditListener.class);
+
+
+  private Map<String, Object> data = Maps.newLinkedHashMap();
+
+  @Override
+  public void onApplicationEvent(AbstractAuthorizationEvent event) {
+
+    data = Maps.newLinkedHashMap();
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthorizationAuditListener.java]: See comment above

File: iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java [hunk 3]
```diff
@@ -42,6 +52,11 @@ public void linkExternalAccount(Principal authenticatedUser,
     IamAccount userAccount = findAccount(authenticatedUser);
 
     externalAuthenticationToken.linkToIamAccount(externalAccountLinker, userAccount);
+
+    eventPublisher.publishEvent(new AccountLinkEvent(this, userAccount,
+        externalAuthenticationToken.toExernalAuthenticationInfo(),
+        String.format("User %s link new %s account", userAccount.getUsername(),
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java]: I would rephrase the message to : "User %s has linked a new account of type %s"

File: iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java [hunk 4]
```diff
@@ -84,6 +99,9 @@ public void unlinkExternalAccount(Principal authenticatedUser, ExternalAuthentic
     if (modified) {
       userAccount.touch();
       iamAccountRepository.save(userAccount);
+
+      eventPublisher.publishEvent(new AccountUnlinkEvent(this, userAccount, type, iss, sub,
+          String.format("User %s unlink %s account", userAccount.getUsername(), type.toString())));
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java]: As above:
User %s has unlinked an account of type %s

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java [hunk 5]
```diff
@@ -139,6 +156,9 @@ public void delete(String id) {
     }
 
     groupRepository.delete(group);
+
+    eventPublisher
+      .publishEvent(new GroupRemoveEvent(this, group, "Deleted group " + group.getName()));
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java]: I would rephrase this: Group %s has been removed.

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java [hunk 6]
```diff
@@ -219,6 +244,9 @@ private void executePatchOperation(IamGroup group, ScimPatchOperation<List<ScimM
         a.touch();
         accountRepository.save(a);
         hasChanged = true;
+
+        eventPublisher.publishEvent(new GroupUpdateEvent(this, group, u.getType(),
+            String.format("Update group %s", group.getName())));
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java]: I would rephrase the message: "Updated information for group %s"

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java [hunk 7]
```diff
@@ -168,6 +188,11 @@ public ScimGroup replace(String id, ScimGroup scimItemToBeReplaced) {
     updatedGroup.touch();
 
     groupRepository.save(updatedGroup);
+
+    eventPublisher
+      .publishEvent(new GroupReplaceEvent(this, updatedGroup, existingGroup, String.format(
+          "Replaced group %s with new group %s", existingGroup.getName(), updatedGroup.getName())));
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java]: I don't like very much this message, but I haven't a good proposal.
It's not much that the group is being replaced, is more some information is being updated.

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java [hunk 8]
```diff
@@ -120,6 +132,8 @@ public void delete(final String id) {
 
     accountRepository.delete(account);
 
+    eventPublisher.publishEvent(
+        new AccountRemoveEvent(this, account, "Delete account for user " + account.getUsername()));
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java]: Delete account for user -> Removed account for user

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java [hunk 9]
```diff
@@ -346,6 +368,9 @@ private void executePatchOperation(IamAccount account, ScimPatchOperation<ScimUs
         throw new ScimPatchOperationNotSupported(u.getType().getDescription() + " not supported");
       }
       hasChanged |= u.update();
+
+      eventPublisher.publishEvent(new AccountUpdateEvent(this, account, u.getType(),
+          String.format("Update account for user %s", account.getUsername())));
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java]: Updated account information for user

File: iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuditField.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package it.infn.mw.iam.audit;
+
+public interface IamAuditField {
+  final String source = "source";
```
Comment (andreaceccanti) [inline iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuditField.java]: Shouldn't these be in UPPER_CASE being constants?

File: iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthenticationAuditListener.java (hunk 1 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthenticationAuditListener.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthorizationAuditListener.java (hunk 2 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuthorizationAuditListener.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java (hunk 3 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java (hunk 4 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/account_linking/DefaultAccountLinkingService.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java (hunk 5 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java (hunk 6 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java (hunk 8 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java (hunk 9 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java]: Done

File: iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuditField.java (hunk 10 above)
Comment (marcocaberletti) [inline iam-login-service/src/main/java/it/infn/mw/iam/audit/IamAuditField.java]: Done
